pkg/controller/volume/vacprotection/vac_protection_controller.go

```diff
@@ -0,0 +1,403 @@
+	vacLister storagelisters.VolumeAttributesClassLister
+	vacSynced cache.InformerSynced
+
+	getPVsAssignedToVAC  func(vacName string) ([]*v1.PersistentVolume, error)
+	getPVCsAssignedToVAC func(vacName string) ([]*v1.PersistentVolumeClaim, error)
+
+	queue workqueue.RateLimitingInterface
+}
+
+// NewVACProtectionController returns a new *Controller.
+func NewVACProtectionController(logger klog.Logger,
+	client clientset.Interface,
+	pvcInformer coreinformers.PersistentVolumeClaimInformer,
+	pvInformer coreinformers.PersistentVolumeInformer,
+	vacInformer storageinformers.VolumeAttributesClassInformer) (*Controller, error) {
+	c := &Controller{
+		client:    client,
+		pvcSynced: pvcInformer.Informer().HasSynced,
+		pvSynced:  pvInformer.Informer().HasSynced,
+		vacLister: vacInformer.Lister(),
+		vacSynced: vacInformer.Informer().HasSynced,
+		queue:     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "vacprotection"),
+	}
+
+	_, _ = vacInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.vacAddedUpdated(logger, obj)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			c.vacAddedUpdated(logger, new)
+		},
+	})
+
+	err := pvInformer.Informer().AddIndexers(cache.Indexers{vacNameKeyIndex: func(obj interface{}) ([]string, error) {
+		pv, ok := obj.(*v1.PersistentVolume)
+		if !ok {
+			return []string{}, nil
+		}
+		return getPVReferencedVACNames(pv), nil
+	}})
+	if err != nil {
+		return nil, fmt.Errorf("failed to add index to PV informer: %w", err)
+	}
+
+	pvIndexer := pvInformer.Informer().GetIndexer()
+	c.getPVsAssignedToVAC = func(vacName string) ([]*v1.PersistentVolume, error) {
+		objs, err := pvIndexer.ByIndex(vacNameKeyIndex, vacName)
+		if err != nil {
+			return nil, err
+		}
+		pvcs := make([]*v1.PersistentVolume, 0, len(objs))
+		for _, obj := range objs {
+			pvc, ok := obj.(*v1.PersistentVolume)
+			if !ok {
+				continue
+			}
+			pvcs = append(pvcs, pvc)
+		}
+		return pvcs, nil
+	}
+
+	err = pvcInformer.Informer().AddIndexers(cache.Indexers{vacNameKeyIndex: func(obj interface{}) ([]string, error) {
+		pvc, ok := obj.(*v1.PersistentVolumeClaim)
+		if !ok {
+			return []string{}, nil
+		}
+		return getPVCReferencedVACNames(pvc), nil
+	}})
+	if err != nil {
+		return nil, fmt.Errorf("failed to add index to PVC informer: %w", err)
+	}
+
+	pvcIndexer := pvcInformer.Informer().GetIndexer()
+	c.getPVCsAssignedToVAC = func(vacName string) ([]*v1.PersistentVolumeClaim, error) {
+		objs, err := pvcIndexer.ByIndex(vacNameKeyIndex, vacName)
+		if err != nil {
+			return nil, err
+		}
+		pvcs := make([]*v1.PersistentVolumeClaim, 0, len(objs))
+		for _, obj := range objs {
+			pvc, ok := obj.(*v1.PersistentVolumeClaim)
+			if !ok {
+				continue
+			}
+			pvcs = append(pvcs, pvc)
+		}
+		return pvcs, nil
+	}
+
+	_, _ = pvcInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old, new interface{}) {
```

Comment:
Do we need to handle Add? Same question with PV handler?

Reply:
Assume that a pvc or pv is created with a given vac.

- If the vac doesn't exist, do thing because we cannot assign a finalizer to the non-existance object.
- If the vac exists, the finalizer should be handled by the event handlers of vac informers in L81.
- If the vac exists but the informer isn't observerd, once the event arrives, the finalizer will be added.

So I don't add AddFunc to pvc/pv informers.